From c77c46540079df61c58991866c28f49b86fe6785 Mon Sep 17 00:00:00 2001
From: Jian-Ming Wang <jian.ming.wang@mitaccomputing.com>
Date: Thu, 16 Apr 2026 08:39:15 +0000
Subject: [PATCH 2/2] Support the association of cooling and cooledby for Fan
 Port.

Description:
1. Example of EM Configuration:

```
{
  "Exposes": [
    {
      "ConnectsToType": "Fan Upstream Port",
      "Name": "Fan 2 Downstream Port",
      "Type": "DownstreamPort",
      "FanPort": true
    },
    {
      "Name": "Fan_2_SPEED_RPM",
      "MaxValue": 20000,
      "MinValue": 0,
      "Sensors": [
        "FCB_Fan_2A_SPEED_RPM",
        "FCB_Fan_2B_SPEED_RPM"
      ],
      "Thresholds": [
        {
          "Direction": "less than",
          "Name": "lower critical",
          "Severity": 1,
          "Value": 1080
        },
        {
          "Direction": "less than",
          "Name": "lower non critical",
          "Severity": 0,
          "Value": 1260
        }
      ],
      "Type": "Minimum",
      "Units": "RPMS"
    }
  ],
  "Name": "Fan 2",
  "Probe": [
    "FOUND('Chassis R520G6')"
  ],
  "Type": "Fan",
  "xyz.openbmc_project.Inventory.Decorator.Asset": {
    "Manufacturer": "Mitac",
    "Model": "FAN6056",
    "PartNumber": "B5678"
  },
  "xyz.openbmc_project.Inventory.Item": {
    "Present": true,
    "PresenceGPIO": {
      "PinName": "FAN2_PRSNT_N",
      "Polarity": "Low"
    }
  },
  "xyz.openbmc_project.Inventory.Item.Fan": {
    "HotPluggable": true,
    "Replaceable": true,
    "RatedSpeedRPM": 12345,
    "RatedSecondarySpeedRPM": 12345,
    "SpeedPercent": "$FCB_Pwm_2_PWM_PCT"
  },
  "xyz.openbmc_project.Inventory.Item.Fan.Oem": {
    "SpeedRPM": "$FCB_Fan_2A_SPEED_RPM",
    "SecondarySpeedRPM": "$FCB_Fan_2B_SPEED_RPM"
  }
}
```

2. Output of Object Mapper

```
busctl --json=pretty get-property xyz.openbmc_project.ObjectMapper /xyz/openbmc_project/inventory/system/chassis/Chassis_R520G6/cooled_by xyz.openbmc_project.Association endpoints
{
        "type" : "as",
        "data" : [
                "/xyz/openbmc_project/inventory/system/fan/Fan_1",
                "/xyz/openbmc_project/inventory/system/fan/Fan_2",
                "/xyz/openbmc_project/inventory/system/fan/Fan_3",
                "/xyz/openbmc_project/inventory/system/fan/Fan_4",
                "/xyz/openbmc_project/inventory/system/fan/Fan_5",
                "/xyz/openbmc_project/inventory/system/fan/Fan_6"
        ]
}

busctl --json=pretty tree xyz.openbmc_project.ObjectMapper
    ...
    |   |- /xyz/openbmc_project/inventory/system/fan
    |   | |- /xyz/openbmc_project/inventory/system/fan/Fan_1
    |   | | |- /xyz/openbmc_project/inventory/system/fan/Fan_1/all_sensors
    |   | | |- /xyz/openbmc_project/inventory/system/fan/Fan_1/contained_by
    |   | | `- /xyz/openbmc_project/inventory/system/fan/Fan_1/cooling
```

Tested:
- R2520G6

Upstream-Status: Inappropriate [No upstream maintenance]
---
 src/entity_manager/topology.cpp | 10 ++++++++++
 src/entity_manager/topology.hpp |  1 +
 2 files changed, 11 insertions(+)

diff --git a/src/entity_manager/topology.cpp b/src/entity_manager/topology.cpp
index fa95a85..ef03f56 100644
--- a/src/entity_manager/topology.cpp
+++ b/src/entity_manager/topology.cpp
@@ -34,6 +34,11 @@ void Topology::addBoard(const std::string& path, const std::string& boardType,
         {
             powerPaths.insert(path);
         }
+        auto findCooledBy = exposesItem.find("FanPort");
+        if (findCooledBy != exposesItem.end())
+        {
+            coolingPaths.insert(path);
+        }
     }
     else if (exposesType.ends_with("Port"))
     {
@@ -75,6 +80,11 @@ std::unordered_map<std::string, std::vector<Association>> Topology::getAssocs(
                             result[upstream].emplace_back(
                                 "powered_by", "powering", downstream);
                         }
+                        if (coolingPaths.contains(downstream))
+                        {
+                            result[upstream].emplace_back(
+                                "cooled_by", "cooling", downstream);
+                        }
                     }
                 }
             }
diff --git a/src/entity_manager/topology.hpp b/src/entity_manager/topology.hpp
index 816704a..7bdeb9f 100644
--- a/src/entity_manager/topology.hpp
+++ b/src/entity_manager/topology.hpp
@@ -28,6 +28,7 @@ class Topology
     std::unordered_map<PortType, std::vector<Path>> upstreamPorts;
     std::unordered_map<PortType, std::vector<Path>> downstreamPorts;
     std::set<Path> powerPaths;
+    std::set<Path> coolingPaths;
     std::unordered_map<Path, BoardType> boardTypes;
     std::unordered_map<BoardName, Path> boardNames;
 };
-- 
2.43.0

